src/universe.rs

```diff
@@ -94,24 +108,96 @@ impl SymbolStore {
         self.name_by_sym.push(None);
         sym
     }
+}
 
+impl Symbols for SymbolStore {
     /// Get the name of a symbol, if it has one.
-    pub fn get_symbol_name(&self, sym: Sym) -> Option<&str> {
+    fn get_symbol_name(&self, sym: Sym) -> Option<&str> {
         self.name_by_sym.get(sym.ord()).and_then(|n| n.as_deref())
     }
 
     /// Returns the number of symbols that have been allocated in this universe.
-    pub fn num_symbols(&self) -> usize {
+    fn num_symbols(&self) -> usize {
         self.name_by_sym.len()
     }
 }
 
+impl<'a> SymbolStorage for SymbolStore {
+    /// Return the symbol associated with the name, or allocate a fresh ID and associate it with the
+    /// given name.
+    fn get_or_insert_named(&mut self, name: &str) -> Sym {
+        self.sym_by_name.get(name).copied().unwrap_or_else(|| {
+            let sym = Sym::from_ord(self.name_by_sym.len());
+            let shared_name: Arc<str> = name.into();
+            self.name_by_sym.push(Some(shared_name.clone()));
+            self.sym_by_name.insert(shared_name, sym);
+            sym
+        })
+    }
+}
+
 impl Default for SymbolStore {
     fn default() -> Self {
         Self::new()
     }
 }
 
+/// Stores and allocates unique symbols on top of a symbol store, without modifying it.
+#[derive(Debug)]
+pub struct SymbolOverlay<'a> {
+    symbols: &'a SymbolStore,
```

Comment:
You don't have to do this change, but I was thinking: theoretically, this `SymbolOverlay` should work with any `symbols: &'a T` where `T: Symbols`, right?

Reply:
Yes, that should be rather straightforward.